src/fasta/faidx.cpp

```diff
@@ -0,0 +1,190 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fstream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "fasta/faidx.hpp"
+
+struct ContigIndex {
+    std::string name;     // Name of this contig
+    uint64_t seq_length;  // Total length of this contig, in bases
+    uint64_t seq_offset;  // Offset in the FASTA file of this contig's first base
+    uint32_t line_length; // The number of bases on each line
+    uint32_t line_bytes;  // The number of bytes in each line, including the newline
+};
+
+class FastaIndexerState;
+
+class FastaIndexer {
+public:
+    std::ostream & index(std::istream & input, std::ostream & output);
+
+private:
+    friend class NewContigState;
+    friend class ContigNameState;
+    friend class ContigDescriptionState;
+    friend class ContigSeqState;
+
+    void set_state(FastaIndexerState* state);
+
+    std::unique_ptr<FastaIndexerState> current_state;
+    std::vector<ContigIndex> fastaIndex;
+};
+
+class FastaIndexerState {
+public:
+    virtual void process(FastaIndexer * fastaIndexer, std::istream & input) = 0;
+    virtual ~FastaIndexerState(){}
+};
+
+class NewContigState : public FastaIndexerState {
+public:
+    void process(FastaIndexer * fastaIndexer, std::istream & input);
+};
+
+class ContigNameState : public FastaIndexerState {
+public:
+    void process(FastaIndexer * fastaIndexer, std::istream & input);
+};
+
+class ContigDescriptionState : public FastaIndexerState {
+public:
+    void process(FastaIndexer * fastaIndexer, std::istream & input);
+};
+
+class ContigSeqState : public FastaIndexerState {
+public:
+    void process(FastaIndexer * fastaIndexer, std::istream & input);
+};
+
+std::ostream &
+FastaIndexer::index(std::istream &input, std::ostream &output) {
+    current_state.reset(nullptr);
+    fastaIndex.clear();
+
+    char c;
+    if (input && input.get(c)) {
+        if (c == '>') {
+            set_state(new NewContigState());
+            while (current_state.get()) {
+                current_state->process(this, input);
+            }
+        }
+    }
+
+    output.clear();
+    for (ContigIndex contigIndex : fastaIndex) {
+        output << contigIndex.name << "\t"
+            << contigIndex.seq_length << "\t"
+            << contigIndex.seq_offset << "\t"
+            << contigIndex.line_length << "\t"
+            << contigIndex.line_bytes << "\n";
+    }
+
+    return output;
+}
+
+void
+FastaIndexer::set_state(FastaIndexerState *state) {
+    current_state.reset(state);
+}
+
+void
+NewContigState::process(FastaIndexer *fastaIndexer, std::istream & input) {
+    fastaIndexer->fastaIndex.push_back({"", 0, 0, 0, 0});
+    fastaIndexer->set_state(new ContigNameState());
+}
+
+void
+ContigNameState::process(FastaIndexer *fastaIndexer, std::istream & input) {
+    char c;
+    ContigIndex & contigIndex = fastaIndexer->fastaIndex[fastaIndexer->fastaIndex.size()-1];
```

Comment:
I would have used 
```suggestion
    ContigIndex & contigIndex = fastaIndexer->fastaIndex.back();
```
but the current version is ok. longer but clearer.

Reply:
I will stick with the original implementation as that's clearer to me.